dart/common/Addon.cpp

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2015, Georgia Tech Research Corporation
+ * All rights reserved.
+ *
+ * Author(s): Michael X. Grey <mxgrey@gatech.edu>
+ *
+ * Georgia Tech Graphics Lab and Humanoid Robotics Lab
+ *
+ * Directed by Prof. C. Karen Liu and Prof. Mike Stilman
+ * <karenliu@cc.gatech.edu> <mstilman@cc.gatech.edu>
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <cassert>
+#include <string>
+#include <iostream>
+
+#include "dart/common/Addon.h"
+#include "dart/common/Console.h"
+
+namespace dart {
+namespace common {
+
+//==============================================================================
+void Addon::setAddonState(const State& /*otherState*/)
+{
+  // Do nothing
+}
+
+//==============================================================================
+const Addon::State* Addon::getAddonState() const
+{
+  return nullptr;
+}
+
+//==============================================================================
+void Addon::setAddonProperties(const Properties& /*someProperties*/)
+{
+  // Do nothing
+}
+
+//==============================================================================
+const Addon::Properties* Addon::getAddonProperties() const
+{
+  return nullptr;
+}
+
+//==============================================================================
+bool Addon::isOptional(AddonManager* /*oldManager*/)
+{
+  return true;
```

Comment:
The current scheme for identifying specialization has zero overhead because specialization is determined at compile-time (notice that my version of ``isSpecializedFor`` is a constexpr). Grabbing a specialized addon is not only constant-time, it is essentially instant. The scheme you're proposing would introduce potentially substantial overhead, especially if the Addon is accessed frequently.

> to using unordered_map for specialized addon map.

That's actually not what I meant to say. I meant to say that an unordered_map should be used to keep track of which addons are **mandatory**, not which ones are specialized. It's important to be clear that specialized addons are not necessarily mandatory, and vice-versa.

MandatoryAddon does not need to be called before SpecializedAddonManager if we store the unordered_map of mandatory addons at the level of the most basic general manager. However, I do agree with having MandatoryAddon inherit SpecializedAddonManager, because there's no reason not to do that.

By the way, I think any of these changes that we're discussing should go into a separate pull request to make it easier to review and evaluate.